File: LC/332/332.py
"""
["JFK","SFO"],["JFK","ATL"],["SFO","ATL"],["ATL","JFK"],["ATL","SFO"]

adjacents = {
    "JFK": ["SFO", "ATL"]
    "SFO": ["ATL"]
    "ATL": ["JFK", "SFO"]
}

 {
     'JFK': [u'ANU', u'TIA'],
     'EZE': [u'AXA'],
     'AXA': [u'TIA'],
     'TIA': [u'ANU', u'ANU', u'JFK'],
     'ANU': [u'JFK', u'EZE', u'TIA']
 }
"""
from collections import defaultdict

class Solution(object):
    def findItinerary(self, tickets):
        """
        :type tickets: List[List[str]]
        :rtype: List[str]
        """
        if not tickets:
            return []
        
        adjacents = defaultdict(lambda: [])
        for f, t in tickets:
            adjacents[f].append(t)
        
        visited = dict()
        for origin, itinerary in adjacents.items():
           itinerary.sort()
            visited[origin] = [False]*len(itinerary)
        
        output = ["JFK"]
        self.dfs(adjacents, "JFK", len(tickets), output, visited)
        return output
        
    
    def dfs(self, adjacents, airport, total_tickets, current, visited):
        if total_tickets + 1 == len(current):
            return True
        
        for i, next_airport in enumerate(adjacents[airport]):
            if not visited[airport][i]:
                current.append(next_airport)
                visited[airport][i] = True
                if self.dfs(adjacents, next_airport, total_tickets, current, visited):
                    return True
                current.pop()
                visited[airport][i] = False
        
        return False
         

File: LC/332/332.java
/*
JFK: [SFO, ATL],
SFO: [ATL],
ATL: [JFK, SFO]

JFK -> SFO -> ATL -> JFK -> ATL -> SFO
JFK -> ATL -> JFK -> SFO -> ATL -> SFO
JFK -> ATL -> SFO -> ATL -> JFK -> SFO

JFK: [SFO, ATL],
SFO: [ATL],
ATL: [JFK]

JFK -> SFO -> ATL -> JFK -> ATL
JFK -> ATL -> JFK -> SFO -> ATL

JFK: [SFO, ATL],
SFO: [],
ATL: [JFK, SFO]
*/
class Solution {
    public List<String> findItinerary(List<List<String>> tickets) {
        if (tickets.size() == 0) return new ArrayList<>();

        List<String> itinerary = new ArrayList<>();
        itinerary.add("JFK");
        Map<String, List<String>> graph = buildGraph(tickets);

        int size = 0;
        Map<String, boolean[]> visitedMap = new HashMap<>();
        for (String from: graph.keySet()) {
            size += graph.get(from).size();
            Collections.sort(graph.get(from));
            visitedMap.put(from, new boolean[graph.get(from).size()]);
        }
        getItinerary("JFK", graph, visitedMap, itinerary, size);
        return itinerary;
    }

    private boolean getItinerary(String from, Map<String, List<String>> graph, Map<String, boolean[]> visitedMap, List<String> itinerary, int size) {
        if (size == 0) return true;

        List<String> tos = graph.getOrDefault(from, new LinkedList<String>());
        boolean[] visited = visitedMap.get(from);

        for (int i = 0; i < tos.size(); i++) {
            if (visited[i]) continue;
            visited[i] = true;
            itinerary.add(tos.get(i));
            if (getItinerary(tos.get(i), graph, visitedMap, itinerary, size-1)) {
                return true;
            }
            itinerary.remove(itinerary.size()-1);
            visited[i] = false;
        }

        return false;
    }

    private Map<String, List<String>> buildGraph(List<List<String>> tickets) {
        Map<String, List<String>> graph = new HashMap<>();
        for (List<String> ticket: tickets) {
            String from = ticket.get(0);
            String to = ticket.get(1);
            if (!graph.containsKey(from)) graph.put(from, new LinkedList<String>());
            graph.get(from).add(to);
        }
        return graph;
    }
}
